src/singleton_snap.py

```diff
@@ -0,0 +1,293 @@
+        """Helper method to construct registration file path.
+
+        The registration file name is constructed as:
+        LCK..{snap_name}--{unit_name}
+        """
+        filename = f"{self.LOCK_FILE_PREFIX}{snap_name}{self.SEPARATOR}{self.unit_name}"
+        return os.path.join(self.LOCK_DIR, filename)
+
+    def _parse_unit_name_from_file(self, filename: str, snap_name: str) -> str:
+        """Helper method to extract unit name from registration filename."""
+        prefix = f"{self.LOCK_FILE_PREFIX}{snap_name}{self.SEPARATOR}"
+        return filename[len(prefix) :]
+
+    def register(self, snap_name: str) -> None:
+        """Register current unit as using the specified snap.
+
+        Raises:
+            SingletonSnapError: if there is an I/O related error creating the lock file.
+        """
+        self._update_registration(snap_name, create=True)
+
+    def unregister(self, snap_name: str) -> None:
+        """Unregister current unit from using the specified snap.
+
+        Raises:
+            SingletonSnapError: if there is an I/O related error removing the lock file.
+        """
+        self._update_registration(snap_name, create=False)
+
+    @contextmanager
+    def _lock_directory(self):
+        dir_fd = None
+        try:
+            dir_fd = os.open(self.LOCK_DIR, os.O_RDONLY)
+            fcntl.flock(dir_fd, fcntl.LOCK_SH)
+            yield
+        except TypeError as e:
+            raise SingletonSnapError(f"Invalid lock directory path: {e}") from e
+        except FileNotFoundError as e:
+            raise SingletonSnapError(f"Lock directory not found: {e}") from e
+        except OSError as e:
+            raise SingletonSnapError(f"Error locking directory: {e}") from e
+        finally:
+            if dir_fd is not None:
+                try:
+                    fcntl.flock(dir_fd, fcntl.LOCK_UN)
+                    os.close(dir_fd)
+                except Exception as e:
+                    raise SingletonSnapError(f"Failed to release lock: {e}") from e
+
+    def _update_registration(self, snap_name: str, create: bool) -> None:
+        """Internal method to handle registration or unregistration of a snap.
+
+        Achieved by creating or removing a lock file; atomic operation is ensured
+        with a directory lock.
+
+        Raises:
+            SingletonSnapError: If an OS error occurs.
+        """
+        lock_path = self._get_registration_file_path(snap_name)
+        action = "registering" if create else "unregistering"
+
+        try:
+            with self._lock_directory():
+                if create:
+                    open(lock_path, "w").close()
+                else:
+                    try:
+                        os.remove(lock_path)
+                    except FileNotFoundError as e:
+                        # pass
+                        raise SingletonSnapError(f"Error {action} unit: {e}") from e
+        except Exception as e:
```

Comment:
This is too broad an exception class to catch. Can we restrict at least to `OSError`? Otherwise it'll catch all manner of things like `NameError` and `AttributeError`.

That makes me think, though. Do users of this module really care about catching `SingletonSnapError`? Maybe we should just let all errors bubble up instead?